pkg/scheduler/apis/config/scheme/scheme.go

```diff
@@ -44,5 +44,5 @@ func AddToScheme(scheme *runtime.Scheme) {
 	utilruntime.Must(kubeschedulerconfigv1.AddToScheme(Scheme))
 	utilruntime.Must(kubeschedulerconfigv1alpha1.AddToScheme(Scheme))
 	utilruntime.Must(kubeschedulerconfigv1alpha2.AddToScheme(Scheme))
-	utilruntime.Must(scheme.SetVersionPriority(kubeschedulerconfigv1alpha1.SchemeGroupVersion))
+	utilruntime.Must(scheme.SetVersionPriority(kubeschedulerconfigv1alpha2.SchemeGroupVersion))
```

Comment:
@liggitt what should we set this to? and what does setting the version priority actually mean?

Reply:
Normally you would specify all known versions here, in the order you would prefer when asked to serialize without specifying a particular version.

This would probably be reasonable:

scheme.SetVersionPriority(kubeschedulerconfigv1alpha2.SchemeGroupVersion, kubeschedulerconfigv1alpha1.SchemeGroupVersion)

Reply:
Uhm... If I remove the version from the config files in the test, it seems to parse, but it doesn't add defaults. So it fails validation next. Any idea what could we be missing?

Reply:
It looks like it's simply not matching any of the versions. And then it goes to the lenient path, in which also it doesn't match.

So the question would be, how do we make our versioned configurations match an unversioned one?
I assume we can do something in https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kube-scheduler/config/v1alpha2/register.go

Reply:
by providing a destination type in the decoding method, it tolerates a missing apiVersion/kind.

I think this will do what you want (require apiVersion/kind, and only handle v1alpha1 in lenient decoding):
```diff
diff --git a/cmd/kube-scheduler/app/options/configfile.go b/cmd/kube-scheduler/app/options/configfile.go
index 0aec0659bff..2c9a0283bb8 100644
--- a/cmd/kube-scheduler/app/options/configfile.go
+++ b/cmd/kube-scheduler/app/options/configfile.go
@@ -41,9 +41,8 @@ func loadConfigFromFile(file string) (*kubeschedulerconfig.KubeSchedulerConfigur
 }
 
 func loadConfig(data []byte) (*kubeschedulerconfig.KubeSchedulerConfiguration, error) {
-       configObj := &kubeschedulerconfig.KubeSchedulerConfiguration{}
        // The UniversalDecoder runs defaulting and returns the internal type by default.
-       err := runtime.DecodeInto(kubeschedulerscheme.Codecs.UniversalDecoder(), data, configObj)
+       obj, _, err := kubeschedulerscheme.Codecs.UniversalDecoder().Decode(data, nil, nil)
        if err != nil {
                // Try strict decoding first. If that fails decode with a lenient
                // decoder, which has only v1alpha1 registered, and log a warning.
@@ -56,17 +55,21 @@ func loadConfig(data []byte) (*kubeschedulerconfig.KubeSchedulerConfiguration, e
                _, lenientCodecs, lenientErr := codec.NewLenientSchemeAndCodecs(
                        kubeschedulerconfig.AddToScheme,
                        kubeschedulerconfigv1alpha1.AddToScheme,
-                       kubeschedulerconfigv1alpha2.AddToScheme,
                )
                if lenientErr != nil {
                        return nil, lenientErr
                }
-               if lenientErr = runtime.DecodeInto(lenientCodecs.UniversalDecoder(), data, configObj); lenientErr != nil {
+               obj, _, lenientErr = lenientCodecs.UniversalDecoder().Decode(data, nil, nil)
+               if lenientErr != nil {
                        return nil, fmt.Errorf("failed lenient decoding: %v", err)
                }
                klog.Warningf("using lenient decoding as strict decoding failed: %v", err)
        }
 
+       configObj, ok := obj.(*kubeschedulerconfig.KubeSchedulerConfiguration)
+       if !ok {
+               return nil, fmt.Errorf("expected KubeSchedulerConfiguration, got %T", obj)
+       }
        return configObj, nil
 }
```

Reply:
I actually wanted to tolerate the missing apiVersion. Or is this a bad practice?

Reply:
We should definitely not tolerate that. Unidentified config content is dangerous to accept.

Reply:
Done. I also updated title and PR description.

Reply:
However, I'm still questioning this line (`SetVersionPriority`). It doesn't seem like version priority is being used anywhere during decoding. And now we are saying that version should be required, so this doesn't make sense.

Reply:
> It doesn't seem like version priority is being used anywhere during decoding

It is used during *encoding* when you don't specify a preferred target version

Reply:
You should specify all known versions in priority order:
```
scheme.SetVersionPriority(kubeschedulerconfigv1alpha2.SchemeGroupVersion, kubeschedulerconfigv1alpha1.SchemeGroupVersion)
```

Reply:
Thanks for all the clarifications. TIL :)